pkg/stanza/fileconsumer/file_threadpool.go

```diff
@@ -0,0 +1,198 @@
+)
+
+type readerWrapper struct {
+	reader *reader
+	fp     *fingerprint.Fingerprint
+}
+
+func (m *Manager) kickoffThreads(ctx context.Context) {
+	m.readerChan = make(chan readerWrapper, m.maxBatchFiles*2)
+	for i := 0; i < m.maxBatchFiles; i++ {
+		m.workerWg.Add(1)
+		go m.worker(ctx)
+	}
+}
+
+func (m *Manager) shutdownThreads() {
+	if m.readerChan != nil {
+		close(m.readerChan)
+	}
+	m.workerWg.Wait()
+	// save off any files left
+	// As we already cancelled our current context, create a new one to save any left offsets
+	// This is only applicable for `filelog.useThreadPool` featuregate
+	ctx, cancel := context.WithCancel(context.Background())
+	m.syncLastPollFilesConcurrent(ctx)
+	cancel()
+}
+
+// poll checks all the watched paths for new entries
+func (m *Manager) pollConcurrent(ctx context.Context) {
+	// Increment the generation on all known readers
+	// This is done here because the next generation is about to start
+	m.knownFilesLock.Lock()
+	for i := 0; i < len(m.knownFiles); i++ {
+		m.knownFiles[i].generation++
+	}
+	m.knownFilesLock.Unlock()
+
+	// Get the list of paths on disk
+	matches, err := m.fileMatcher.MatchFiles()
+	if err != nil {
+		m.Errorf("error finding files: %s", err)
+	}
+	m.consumeConcurrent(ctx, matches)
+	m.clearCurrentFingerprints()
+
+	// Any new files that appear should be consumed entirely
+	m.readerFactory.fromBeginning = true
+	m.syncLastPollFilesConcurrent(ctx)
+}
+
+func (m *Manager) worker(ctx context.Context) {
+	defer m.workerWg.Done()
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case chanData, ok := <-m.readerChan:
+			if !ok {
+				return
+			}
+			r, fp := chanData.reader, chanData.fp
+			if !m.readToEnd(ctx, r) {
+				// Save off any files that were not fully read or if deleteAfterRead is disabled
+				m.knownFilesLock.Lock()
+				m.knownFiles = append(m.knownFiles, r)
+				m.knownFilesLock.Unlock()
+			}
+			m.removePath(fp)
+		}
+	}
+}
+
+func (m *Manager) makeReaderConcurrent(filePath string) (*reader, *fingerprint.Fingerprint) {
+	fp, file := m.makeFingerprint(filePath)
+	if fp == nil {
+		return nil, nil
+	}
+
+	// check if the current file is already being consumed
+	if m.isCurrentlyConsuming(fp) {
+		if err := file.Close(); err != nil {
+			m.Errorf("problem closing file", "file", file.Name())
+		}
+		return nil, nil
+	}
+
+	// Exclude any empty fingerprints or duplicate fingerprints to avoid doubling up on copy-truncate files
+	if m.checkDuplicates(fp) {
+		if err := file.Close(); err != nil {
+			m.Errorf("problem closing file", "file", file.Name())
+		}
+		return nil, nil
+	}
+	m.currentFps = append(m.currentFps, fp)
+
+	reader, err := m.newReaderConcurrent(file, fp)
+	if err != nil {
+		m.Errorw("Failed to create reader", zap.Error(err))
+		return nil, nil
+	}
+	return reader, fp
+}
+
+func (m *Manager) consumeConcurrent(ctx context.Context, paths []string) {
+	m.clearOldReadersConcurrent(ctx)
```

Comment:
This appears to be a blocking call, so doesn't it effectively segments all file reading into poll intervals? I think we may still have performance benefits from reading files asynchronously _within_ a poll interval (as opposed to processing all files in a batch before moving to the next batch).

Reply:
Okay, I have changed it to non-blocking. 
i just submit it to the channels, just like others and update the trie